File: backend/models/order.py
from mongoengine import Document, StringField, FloatField, IntField, ListField, EmbeddedDocument, EmbeddedDocumentField, ReferenceField, DateTimeField
import datetime
from .user import User
from .dish import Dish

class OrderItem(EmbeddedDocument):
    dish = ReferenceField(Dish, required=True)
    quantity = IntField(required=True, min_value=1)
    price = FloatField(required=True)

class Order(Document):
    customer = ReferenceField(User, required=True)
    items = ListField(EmbeddedDocumentField(OrderItem), default=[])

    original_price = FloatField(default=0.0) # total before discounts for VIP users
    discount_applied = FloatField(default=0.0) # discount amount for VIP users
    final_price = FloatField(default=0.0) # total after discounts

    status = StringField(
        required=True,
        choices=["PendingPayment",
                 "Paid",
                 "Rejected_Insufficient_Funds",
                 "Queued_For_Preparation",
                 "In_Preparation",
                 "On_Hold",
                 "Ready_For_Delivery",
                 "Awaiting_Pickup",
                 "Out_For_Delivery",
                 "Completed",
                 "Delivery_Failed"
                 ],
        default="PendingPayment"
    )
    notes = StringField(default="")
    created_at = DateTimeField(default=datetime.datetime.utcnow)
    updated_at = DateTimeField(default=datetime.datetime.utcnow)

    meta = {
        "collection": "orders",
        "indexes": [
            "customer",
            "status",
            "created_at",
        ]
    }
    def calculate_total_price(self):
        total = sum(item.price * item.quantity for item in self.items)
        self.original_price = total
        if self.customer.isVIP:
            self.discount_applied = round(total * 0.1, 2)  # 10% discount for VIPs
        else:
            self.discount_applied = 0.0

        self.final_price = total - self.discount_applied
        self.save()
    
    def set_status(self, new_status):
        self.status = new_status
        self.updated_at = datetime.datetime.utcnow()
        self.save()
    
    def add_note(self, text):
        self.notes = text
        self.save()
    
    def increment_dish_order_counts(self):
        for item in self.items:
            item.dish.increment_order_count()

File: backend/models/Order.js
const mongoose = require('mongoose');

const orderItemSchema = new mongoose.Schema({
  dish: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Dish',
    required: true
  },
  quantity: {
    type: Number,
    required: true,
    min: 1
  },
  price: {
    type: Number,
    required: true
  }
}, { _id: false });

const orderSchema = new mongoose.Schema({
  customer: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  items: [orderItemSchema],

  original_price: {
    type: Number,
    default: 0.0
  },
  discount_applied: {
    type: Number,
    default: 0.0
  },
  final_price: {
    type: Number,
    default: 0.0
  },

  status: {
    type: String,
    required: true,
    enum: [
      "PendingPayment",
      "Paid",
      "Rejected_Insufficient_Funds",
      "Queued_For_Preparation",
      "In_Preparation",
      "On_Hold",
      "Ready_For_Delivery",
      "Awaiting_Pickup",
      "Out_For_Delivery",
      "Completed",
      "Delivery_Failed"
    ],
    default: "PendingPayment"
  },
  notes: {
    type: String,
    default: ""
  },
  created_at: {
    type: Date,
    default: Date.now
  },
  updated_at: {
    type: Date,
    default: Date.now
  }
}, {
  collection: 'orders'
});

// indexes
orderSchema.index({ customer: 1 });
orderSchema.index({ status: 1 });
orderSchema.index({ created_at: 1 });

// methods
orderSchema.methods.calculate_total_price = async function() {
  let total = 0;
  for (const item of this.items) {
    total += item.price * item.quantity;
  }
  
  this.original_price = total;
  
  // populate customer to check vip status if not already populated
  if (!this.customer || typeof this.customer.isVIP === 'undefined') {
    await this.populate('customer');
  }
  
  if (this.customer.isVIP) {
    this.discount_applied = Math.round(total * 0.1 * 100) / 100; // 10% discount
  } else {
    this.discount_applied = 0.0;
  }

  this.final_price = total - this.discount_applied;
  return this.save();
};

orderSchema.methods.set_status = function(new_status) {
  this.status = new_status;
  this.updated_at = new Date();
  return this.save();
};

orderSchema.methods.add_note = function(text) {
  this.notes = text;
  return this.save();
};

orderSchema.methods.increment_dish_order_counts = async function() {
  const Dish = mongoose.model('Dish');
  for (const item of this.items) {
    await Dish.findByIdAndUpdate(item.dish, { $inc: { order_count: 1 } });
  }
};

module.exports = mongoose.model('Order', orderSchema);
